File: z2.cpp
//var 3
#include <iostream>

using namespace std;

int cost(int v, int n, int *h, int *c, int currentC, int currentV)
{
    if(currentV > v)
        return -1;
    if(n == 0)
        return currentC;
        
    int max = currentC;
    for(int i = 0; i < n; i++)
    {
        int t = cost(v, n-i-1, &h[i+1], &c[i+1], currentC + c[i], currentV + h[i]);
        if(t > max)
        {
            max = t;
        }
    }
    return max;
}

int main()
{
    cout << "Введите грузоподъемность и количество предметов\n";
    int v,n;
    cin >> v >> n;
    cout << "Введите вес предметов\n";
    int h[n];
    for(int i = 0; i < n;i++)
        cin >> h[i];
    cout << "Введите стоимость предметов\n";
    int c[n];
    for(int i = 0; i < n;i++)
        cin >> c[i];
    cout << cost(v, n, h, c, 0, 0);
    return 0;
}

File: z2.py
def cost(v, n, h, c, currentC, currentV, offset):
    if currentV > v:
        return -1
    if offset == n:
        return currentC
    
    max = currentC
    for i in range(offset, n):
        t = cost(v,n,h,c, currentC + c[i], currentV + h[i], i+1)
        if t > max:
            max = t

    return max

print("Введите грузоподъемность и количество предметов")
v = int(input())
n = int(input())
print("Введите вес предметов")
h = []
for i in range(n):
    h.append(int(input()))
print("Введите стоимость предметов")
c = []
for i in range(n):
    c.append(int(input()))

print(cost(v,n,h,c,0,0,0))